text/3537-msrv-resolver.md

```diff
@@ -0,0 +1,1238 @@
+
+After some time, I run:
+```console
+$ cargo update
+Name          Current Latest Note
+============= ======= ====== ==================
+clap          5.10.30 5.11.0 requires Rust 1.93
+clap_complete 5.10.40 5.11.0 requires Rust 1.93
+note: To use the latest depednencies, run `rustup update && cargo update`.
+To force the use of the latest dependencies, independent of your toolchain, run `cargo update --ignore-rust-version`
+```
+At this point, 1.95 is out, so I'm fine updating my MSRV and I run:
+```console
+$ cargo update --update-rust-version
+Updating clap 5.10.30 to 5.11.0
+Updating foo's rust-version from 1.92 to 1.93
+```
+Instead, if a newer clap version was out needing 1.94 or 1.95, I would instead edit `Cargo.toml` myself.
+
+</details>
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We expect these changes to be independent enough and beneficial on there own that they can be stabilized as each is completed.
+
+## Cargo Resolver
+
+We will be adding a v3 resolver, specified through `workspace.resolver` / `package.resolver`.
+This will become default with the next Edition.
+
+When `resolver = "3"` is set, Cargo's resolver will change to *prefer* MSRV compatible versions over
+incompatible versions when resolving new dependencies, except for `cargo install`.
+Initially, dependencies without `package.rust-version` will be preferred over
+MSRV-incompatible packages but less than those that are compatible.
+The exact details for how preferences are determined may change over time,
+particularly when no MSRV is specified,
+but this shouldn't affect existing `Cargo.lock` files since the currently
+resolved dependencies always get preference.
+
+This can be overridden with `--ignore-rust-version` and config's `resolver.precedence`.
+
+Implications
+- If you use `cargo update --precise <msrv-incompatible-ver>`, it will work
+- If you use `--ignore-rust-version` once, you don't need to specify it again to keep those dependencies though you might need it again on the next edit of `Cargo.toml` or `cargo update` run
+- If a dependency doesn't specify `package.rust-version` but its transitive dependencies specify an incompatible `package.rust-version`,
+  we won't backtrack to older versions of the dependency to find one with a MSRV-compatible transitive dependency.
+- A package with multiple MSRVs, depending on the features selected, can still do this as version requirements can still require versions newer than the MSRV and `Cargo.lock` can depend on those as well.
+
+As there is no `workspace.rust-version`,
+the resolver will pick the lowest version among workspace members.
+This will be less optimal for workspaces with multiple MSRVs and dependencies unique to the higher-MSRV packages.
+Users can workaround this by raising the version requirement or using `cargo update --precise`.
+
+When `rust-version` is unset,
+we'll fallback to `rustc --version`.
```

Comment:
Is it only the major.minor.patch portion of the output?

Reply:
I've updated this to specify that we will only do this for stable versions but left it as an unresolved question for is to resolve during stabilization because we can look into doing things like `<major>.<minor> - 1.<patch>`.